usersync/bidderchooser.go

```diff
@@ -0,0 +1,60 @@
+package usersync
+
+// bidderChooser determines which bidders to consider for user syncing.
+type bidderChooser interface {
+	// choose returns an ordered collection of potentially non-unique bidders.
+	choose(requested, available []string, cooperative Cooperative) []string
+}
+
+// standardBidderChooser implements the bidder choosing algorithm per official Prebid specification.
+type standardBidderChooser struct {
+	shuffler shuffler
+}
+
+func (c standardBidderChooser) choose(requested, available []string, cooperative Cooperative) []string {
+	if len(requested) == 0 {
+		return c.shuffledCopy(available)
+	}
```

Comment:
It seems like it would make sense to prioritize the priority groups if cooperative is enabled, but requested was empty. If we call `chooseCooperative()` before checking for an empty `requested`, we will want to make sure `shuffledAppend()` deals well with an empty append. Which may not be a bad thing to verify in general.

Reply:
The prebid.org docs don't have cooperative mode documented yet, and issue #924 doesn't seem to mention the possible case. My best guess is that `requested` is intended to be the bidders on the page, and the priority groups are prioritizing any other bidders to be synced in their groups. In that world, you would only get an empty `requested` if the page had no prebid ads, which seems like a valid, if rare, use case. And of course there is the case where the publisher doesn't follow the (nonexistent) docs correctly, and only fills in the priority groups, which makes sense as a way to go. (what if you want to ensure some bidders on the page are synced before others.)

Reply:
I agree with the proposal to consider cooperative priority groups if no bidders are specified. The end result is the same where all bidders are equally considered if priority groups aren't specified. Changed.

> we will want to make sure shuffledAppend() deals well with an empty append
There is already test coverage for this case.

Reply:
This change looks good. One more possible case to consider is if cooperative cookie sync is off but the limit specified is greater than the number of bidders requested. Right now we are just considering the bidders that were requested but should we also consider other available enabled bidders to reach the limit?
For example, cooperative sync is off, the request limit is five but only two bidders are requested. Should we also sync three other bidders?

Reply:
> Right now we are just considering the bidders that were requested

Correct. This is the desired behavior.

> should we also consider other available enabled bidders to reach the limit?

We should not. That's not how the endpoint is designed to work.